File: solutions/lua/affine-cipher/1/affine-cipher.lua
local function gcd(a, b)
  while b ~= 0 do a, b = b, a % b end
  return a
end

local function inv(n)
  for x = 1, 25 do
    if (x * n) % 26 == 1 then return x end
  end
end

return {
  encode = function(phrase, key)
    assert(gcd(key.a, 26) == 1, "a and m must be coprime.")
    return phrase:lower():gsub("[^%da-z]", ""):gsub("%l", function(ch)
          return string.char(97 + (key.a * (ch:byte() - 97) + key.b) % 26)
        end)
        :gsub("(.....)", "%1 "):gsub("%s+$", "")
  end,

  decode = function(phrase, key)
    assert(gcd(key.a, 26) == 1, "a and m must be coprime.")
    local ainv = inv(key.a)
    return phrase:gsub("[^%da-z]", ""):gsub("%l", function(ch)
      return string.char(97 + ainv * (ch:byte() - 97 - key.b) % 26)
    end)
  end
}

File: solutions/ruby/affine-cipher/1/affine_cipher.rb
# frozen_string_literal: true

# class Affine
class Affine
  def initialize(key_a, key_b)
    raise ArgumentError if key_a.gcd(26) != 1

    @key = Struct.new(:a, :b).new(key_a, key_b)
  end

  def encode(phrase)
    phrase.downcase
          .gsub(/\W/, '')
          .gsub(/[a-z]/) { (97 + (@key.a * (_1.ord - 97) + @key.b) % 26).chr }
          .scan(/.{1,5}/)
          .join(' ')
  end

  def decode(phrase)
    ainv = 1.upto(25).find { (_1 * @key.a) % 26 == 1 }
    phrase.gsub(/\W/, '')
          .gsub(/[a-z]/) { (97 + ainv * (_1.ord - 97 - @key.b) % 26).chr }
  end
end
